pkg/datasource/columns.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datasource
+
+import (
+	"fmt"
+	"reflect"
+	"slices"
+	"unsafe"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns/formatter/json"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/parser"
+)
+
+type DataTuple struct {
+	ds   DataSource
+	data *data
+}
+
+func NewDataTuple(ds DataSource, d Data) *DataTuple {
+	return &DataTuple{
+		ds:   ds,
+		data: d.(*data),
+	}
+}
+
+func (ds *dataSource) Parser() (parser.Parser, error) {
+	cols, err := ds.Columns()
+	if err != nil {
+		return nil, err
+	}
+	return parser.NewParser(cols), nil
+}
+
+func (ds *dataSource) JSONFormatter() (*json.Formatter[DataTuple], error) {
+	cols, err := ds.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("creating columns: %w", err)
+	}
+	formatter := json.NewFormatter(cols.GetColumnMap())
+	return formatter, nil
+}
+
+func (ds *dataSource) Columns() (*columns.Columns[DataTuple], error) {
+	cols, err := columns.NewColumns[DataTuple]()
+	if err != nil {
+		return nil, err
+	}
+
+	for i, f := range ds.fields {
+		if FieldFlagEmpty.In(f.Flags) {
+			continue
+		}
+		if f.ReflectType() == nil {
+			df := columns.DynamicField{
+				Attributes: &columns.Attributes{
+					Name:    f.Name,
```

Comment:
Shouldn't it be f.FullName? Otherwise the DNS gadget fails with:

```
DEBU[0000] failed to get parser: %!w(*fmt.wrapError=&{creating columns: duplicate column name "string" 0xc003fad770})
```

But, if I change that to f.FullName, the trace_tcp gadget leaks a lot of memory and is killed by the OS.

Reply:
Solved 😄